File: dotnet/src/Envdoctor/Commands/SyncCommand.cs
using Envdoctor.Config;
using Envdoctor.Core;
using Envdoctor.Models;

namespace Envdoctor.Commands;

public sealed class SyncArgs
{
    public string? Root { get; set; }
    public string From { get; set; } = "";
    public string To { get; set; } = "";
    public bool DryRun { get; set; }
}

public static class SyncCommand
{
    /// `envdoctor sync <from> <to>` — copy missing variable keys from one
    /// environment file to another, using placeholder values. Never copies
    /// real secret values.
    public static int Run(SyncArgs args)
    {
        var root = Path.GetFullPath(args.Root ?? ".");

        var fromLabel = Shared.NormalizeEnvLabel(args.From);
        var toLabel = Shared.NormalizeEnvLabel(args.To);

        var (model, config) = Pipeline.LoadProject(root);

        var fromNames = NamesForEnvironment(model, fromLabel);
        var toNames = NamesForEnvironment(model, toLabel);

        var missing = fromNames.Where(n => !toNames.Contains(n)).ToList();

        if (missing.Count == 0)
        {
            Console.Out.WriteLine($"✓ {fromLabel} → {toLabel}: nothing to sync");
            return 0;
        }

        var targetFile = TargetEnvPath(root, toLabel, config);
        var targetRel = Discover.RelativeTo(root, targetFile) ?? targetFile;

        var lines = new List<string> { "", $"# Synced from {fromLabel} by envdoctor" };
        foreach (var name in missing)
        {
            var placeholder = EnvironmentVariable.IsSecretName(name) ? "" : $"your_{name.ToLowerInvariant()}";
            lines.Add($"{name}={placeholder}");
        }
        var append = string.Join("\n", lines) + "\n";

        if (args.DryRun)
        {
            Console.Out.WriteLine("envdoctor sync (dry run)\n");
            Console.Out.WriteLine($"Would append {missing.Count} key{Plural(missing.Count)} to {targetRel}:");
            foreach (var name in missing)
                Console.Out.WriteLine($"  + {name}");
            return 0;
        }

        var parentDir = Path.GetDirectoryName(targetFile);
        if (!string.IsNullOrEmpty(parentDir))
            Directory.CreateDirectory(parentDir);
        File.AppendAllText(targetFile, append);

        Console.Out.WriteLine("envdoctor sync\n");
        Console.Out.WriteLine($"  ✓ Appended {missing.Count} key{Plural(missing.Count)} to {targetRel}");
        foreach (var name in missing)
            Console.Out.WriteLine($"    + {name}");

        return 0;
    }

    /// Sorted set of variable names defined for an environment label.
    private static SortedSet<string> NamesForEnvironment(ProjectModel model, string label)
    {
        var names = new SortedSet<string>(StringComparer.Ordinal);
        foreach (var file in model.EnvFiles)
        {
            if (file.Environment == label)
            {
                foreach (var v in file.Variables)
                    names.Add(v.Name);
            }
        }
        return names;
    }

    /// Resolve the file that should receive synced keys for a target label.
    private static string TargetEnvPath(string root, string label, EnvdoctorConfig config)
    {
        if (config.Environments is not null &&
            config.Environments.TryGetValue(label, out var files) &&
            files.Count > 0)
        {
            return Path.Combine(root, files[0]);
        }
        return label == "development"
            ? Path.Combine(root, ".env")
            : Path.Combine(root, $".env.{label}");
    }

    private static string Plural(int n) => n == 1 ? "" : "s";
}

File: swift/Sources/EnvdoctorKit/Commands/SyncCommand.swift
import Foundation

public struct SyncArgs {
    public var root: String?
    public var from: String
    public var to: String
    public var dryRun = false

    public init(from: String, to: String) {
        self.from = from
        self.to = to
    }
}

public enum SyncCommand {
    /// `envdoctor sync <from> <to>` — copy missing variable keys from one
    /// environment file to another, using placeholder values. Never copies
    /// real secret values.
    public static func run(_ args: SyncArgs) throws -> Int {
        let root = resolveRoot(args.root)
        let fromLabel = normalizeEnvLabel(args.from)
        let toLabel = normalizeEnvLabel(args.to)

        let context = try Pipeline.loadProject(root)
        let model = context.model

        let fromNames = namesForEnvironment(model.envFiles, fromLabel)
        let toNames = namesForEnvironment(model.envFiles, toLabel)

        // The reference sorts with the default (code-unit) comparator.
        let missing = fromNames.filter { !toNames.contains($0) }.sorted()

        if missing.isEmpty {
            print("✓ \(fromLabel) → \(toLabel): nothing to sync")
            return 0
        }

        let targetFile = targetEnvPath(root, toLabel, context.config)
        let targetRel = relativePath(root, targetFile)

        var lines: [String] = ["", "# Synced from \(fromLabel) by envdoctor"]
        for name in missing {
            let placeholder = isSecretName(name) ? "" : "your_\(name.lowercased())"
            lines.append("\(name)=\(placeholder)")
        }
        let append = lines.joined(separator: "\n") + "\n"

        if args.dryRun {
            print("envdoctor sync (dry run)\n")
            print("Would append \(missing.count) key\(missing.count == 1 ? "" : "s") to \(targetRel):")
            for name in missing {
                print("  + \(name)")
            }
            return 0
        }

        let dir = (targetFile as NSString).deletingLastPathComponent
        try FileManager.default.createDirectory(atPath: dir, withIntermediateDirectories: true)
        let handle = try FileHandle(forWritingTo: URL(fileURLWithPath: targetFile))
        handle.seekToEndOfFile()
        handle.write(Data(append.utf8))
        try handle.close()

        print("envdoctor sync\n")
        print("  ✓ Appended \(missing.count) key\(missing.count == 1 ? "" : "s") to \(targetRel)")
        for name in missing {
            print("    + \(name)")
        }

        return 0
    }

    private static func namesForEnvironment(_ envFiles: [EnvironmentFile], _ label: String) -> Set<String> {
        var names = Set<String>()
        for file in envFiles where file.environment == label {
            for v in file.variables { names.insert(v.name) }
        }
        return names
    }

    private static func targetEnvPath(_ root: String, _ label: String, _ config: EnvdoctorConfig) -> String {
        if let files = config.environments?[label], let first = files.first {
            return resolvePath(root, first)
        }
        if label == "development" { return root + "/.env" }
        return root + "/.env.\(label)"
    }
}
